src/WaveParallel.cpp

```diff
@@ -0,0 +1,510 @@
+        solution_u_owned.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+
+        solution_v.reinit(locally_owned_dofs, locally_relevant_dofs, MPI_COMM_WORLD);
+        solution_v_owned.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+
+        old_solution_u.reinit(locally_owned_dofs, locally_relevant_dofs, MPI_COMM_WORLD);
+        old_solution_u_owned.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+
+        old_solution_v.reinit(locally_owned_dofs, locally_relevant_dofs, MPI_COMM_WORLD);
+        old_solution_v_owned.reinit(locally_owned_dofs, MPI_COMM_WORLD);
+    }
+    timer.exit_section("Setup");
+pcout << " ======================================== " << std::endl;
+
+}
+
+template <unsigned int dim>
+void WaveEquationParallel<dim>::assemble_matrices()
+{
+
+    timer.enter_section("Assemble Mass/Laplace");
+
+    pcout << " WAVEPARALLEL - ASSEMBLING MATRICES." << std::endl;
+    pcout << " ======================================== " << std::endl;
+
+    const unsigned int dofs_per_cell = fe->dofs_per_cell;
+    const unsigned int quadrature_points = quadrature->size();
+
+    FEValues fe_values(
+        *fe,
+        *quadrature,
+        update_values | update_gradients | update_JxW_values | update_quadrature_points
+    );
+
+    FullMatrix<double> cell_mass_matrix(dofs_per_cell);
+    FullMatrix<double> cell_laplace_matrix(dofs_per_cell);
+
+    std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+    mass_matrix = 0.0;
+    laplace_matrix = 0.0;
+
+    for (const auto& cell : dof_handler.active_cell_iterators())
+    {
+        if (!cell->is_locally_owned())
+            continue;
+
+        fe_values.reinit(cell);
+
+        cell_mass_matrix = 0.0;
+        cell_laplace_matrix = 0.0;
+
+        for (unsigned int q = 0; q < quadrature_points; ++q)
+        {
+            for (unsigned int i = 0; i < dofs_per_cell; ++i)
+            {
+                for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                {
+                    cell_mass_matrix(i, j) +=
+                        fe_values.shape_value(i, q) *
+                        fe_values.shape_value(j, q) *
+                        fe_values.JxW(q);
+
+                    cell_laplace_matrix(i, j) +=
+                        fe_values.shape_grad(i, q) *
+                        fe_values.shape_grad(j, q) *
+                        fe_values.JxW(q);
+                }
+            }
+        }
+
+        cell->get_dof_indices(local_dof_indices);
+        mass_matrix.add(local_dof_indices, cell_mass_matrix);
+        laplace_matrix.add(local_dof_indices, cell_laplace_matrix);
+    }
+    mass_matrix.compress(VectorOperation::add);
+    laplace_matrix.compress(VectorOperation::add);
+
+    timer.exit_section("Assemble Mass/Laplace");
+}
+
+template <unsigned int dim>
+void WaveEquationParallel<dim>::run()
+{
+    timer.enter_section("Run");
+    timer.enter_subsection("Init");
```

Comment:
This can be made simpler and more C++-ish by using RAII through `TimerOutput::Scope`.